src/nbl/asset/utils/ISPIRVDebloater.cpp

```diff
@@ -45,71 +114,131 @@ ISPIRVDebloater::Result ISPIRVDebloater::debloat(const  ICPUBuffer* spirvBuffer,
     while (offset < spirvDwordCount) {
         const auto curOffset = offset;
         const auto instruction = spirv[curOffset];
+        const auto [length, opcode] = parse_instruction(instruction);
+        if (opcode != spv::OpEntryPoint) break;
+        offset += length;
 
-        const auto length = instruction >> 16;
-        const auto opcode = instruction & 0x0ffffu;
+        const auto curExecutionModel = static_cast<spv::ExecutionModel>(spirv[curOffset + 1]);
+        const auto curEntryPointId = spirv[curOffset + 2];
+        const auto curEntryPointName = std::string_view(reinterpret_cast<const char*>(spirv + curOffset + 3));
 
-        offset += length;
+        const auto entryPoint = EntryPoint{
+            .name = curEntryPointName,
+            .shaderStage = getHlslShaderStage(curExecutionModel),
+        };
 
-        if (opcode == spv::OpEntryPoint)
+        if (entryPoint.shaderStage == hlsl::ESS_UNKNOWN)
         {
-            const auto curExecutionModel = static_cast<spv::ExecutionModel>(spirv[curOffset + 1]);
-
-            // TODO: check whether this reinterpret_cast is UB
-            const auto curEntryPointName = std::string_view(reinterpret_cast<const char*>(spirv + curOffset + 3)); // entryPoint name is the third parameter
+            logger.log("Found entry point with unsupported execution model in SPIR-V", system::ILogger::ELL_ERROR);
+            return Result{
+                .spirv = nullptr,
+                .isSuccess = false
+            };
+        }
 
-            auto findEntryPointIt = std::find(entryPoints.begin(), entryPoints.end(), 
-              EntryPoint{
-                .name = curEntryPointName,
-                .shaderStage = getHlslShaderStage(curExecutionModel),
-              });
-            if (findEntryPointIt != entryPoints.end())
-            {
-                foundEntryPoints[std::distance(findEntryPointIt, entryPoints.begin())] = true;
-            } else
+        auto findEntryPointIt = unfoundEntryPoints.find(entryPoint);
+        if (findEntryPointIt != unfoundEntryPoints.end())
+        {
+            unfoundEntryPoints.erase(findEntryPointIt);
+        } else
+        {
+            if (needDebloat == false)
             {
-                if (needDebloat == false)
-                {
-                    minimizedSpirv.reserve(spirvDwordCount);
-                    minimizedSpirv.insert(minimizedSpirv.end(), spirv, spirv + curOffset);
-                    needDebloat = true;
-                }
-                continue;
+                minimizedSpirv.reserve(spirvDwordCount);
+                minimizedSpirv.insert(minimizedSpirv.end(), spirv, spirv + curOffset);
+                needDebloat = true;
             }
+            removedEntryPointIds.insert(curEntryPointId);
+            continue;
         }
-        
         if (!needDebloat) continue;
         minimizedSpirv.insert(minimizedSpirv.end(), spirv + curOffset, spirv + offset);
```

Comment:
because of the `break` on the first instruction that's not an entry point, your `minimizedSpirv` will only contain entry point opcodes !